test/Table_Tests/src/Common_Table_Operations/Conversion_Spec.enso

```diff
@@ -235,6 +235,72 @@ add_specs suite_builder setup =
+                td.at "bools" . value_type . should_equal (Value_Type.Decimal Nothing Nothing)
+                td.at "decimals" . value_type . should_equal (Value_Type.Decimal Nothing Nothing)
+                td.at "ints" . to_vector . should_equal [Decimal.new 1, Decimal.new 2, Decimal.new 4]
+                td.at "floats" . to_vector . should_equal [Decimal.new "1.0001", Decimal.new "2", Decimal.new "4.0"]
+                td.at "bools" . to_vector . should_equal [Decimal.new "1", Decimal.new "2", Decimal.new "1"]
+                td.at "decimals" . to_vector . should_equal [Decimal.new "3", Decimal.new "9223372036854775807002.3", Decimal.new "9223372036854775807003"]
+
+            group_builder.specify "Can cast from Decimal in mixed column" <|
+                t = table_builder [["ints", [Decimal.new 1, Decimal.new 2, 4]], ["floats", [1.0001, Decimal.new "2.25", Decimal.new "4.0"]], ["decimals", [3, Decimal.new "9223372036854775807002.3", Decimal.new "9223372036854775807003"]]]
+                td = t.cast ["ints"] Value_Type.Integer . cast ["floats"] Value_Type.Float . cast ["decimals"] Value_Type.Decimal
+                td.at "ints" . value_type . should_equal Value_Type.Integer
+                td.at "floats" . value_type . should_equal Value_Type.Float
+                td.at "decimals" . value_type . should_equal Value_Type.Decimal
+                td.at "ints" . to_vector . should_equal [1, 2, 4]
+                td.at "floats" . to_vector . should_equal [1.0001, 2.25, 4.0]
+                td.at "decimals" . to_vector . should_equal [Decimal.new "3", Decimal.new "9223372036854775807002.3", Decimal.new "9223372036854775807003"]
+
+            group_builder.specify "Attaches warning on conversion failure" <|
+                t16 = table_builder [["x", [Decimal.new "65539"]]] . cast ["x"] (Value_Type.Integer Bits.Bits_16)
+                Problems.expect_only_warning Conversion_Failure t16
+                t32 = table_builder [["x", [Decimal.new "4294967298"]]] . cast ["x"] (Value_Type.Integer Bits.Bits_32)
+                Problems.expect_only_warning Conversion_Failure t32
+                t64 = table_builder [["x", [Decimal.new "9223372036854775809"]]] . cast ["x"] (Value_Type.Integer Bits.Bits_64)
+                Problems.expect_only_warning Conversion_Failure t64
+
+            group_builder.specify "Attaches warning on loss of precision" <|
+                t0 = table_builder [["x", [Decimal.new "0.1"]]] . cast ["x"] Value_Type.Float
+                Problems.expect_only_warning Loss_Of_Decimal_Precision t0
+                t1 = table_builder [["x", [Decimal.new "1E310"]]] . cast ["x"] Value_Type.Float
+                Problems.expect_only_warning Loss_Of_Decimal_Precision t1
+                t2 = table_builder [["x", [Decimal.new "9223372036854775807000"]]] . cast ["x"] Value_Type.Float
+                Problems.expect_only_warning Loss_Of_Decimal_Precision t2
```

Comment:
I'm wondering if we should be attaching the warning in case of explicit cast.

We do have the `Loss_Of_Integer_Precision` but IMO that is warranted - it is not completely intuitive for most users that casting an int to float can make it look like a different integer - this is unexpected and happens only for very large values - so it is good to warn the user that an unusual thing is happening.

The `Loss_Of_Decimal_Precision` will be triggered by almost every number. It is expected that converting from _exact_ Decimal type to an inexact `Float` type will lose precision. I'm not really sure that the warning here is needed. It will happen all the time so it will just get ignored and won't serve much purpose.

What do you think? I'm also okay with keeping it, but IMHO it may be better to avoid it when casting explicitly.

Reply:
Another story is probably in cases where we have a `BigDecimal` value that is so big that converting it, even the integral part will change - i.e. it's not just a `0.2999999999` changing to `0.3` or vice versa, but it is `123456789000000000123.000001` changing into `123456789000000000000.0`.

But even here, I think it is overall expected that `Decimal` -> `Float` loses precision, so I think _not warning when explicitly casting_ is OK.

Reply:
The case of `Integer` -> `Float` is special in the way that most users will expect the other way round: `Float` -> `Integer` to be losing precision - because we lose the fractional part; but then the reverse `Integer` -> `Float` is generally understood to be a conversion into a 'wider' type and is often (incorrectly) assumed to be exact. And this will not show for most practical values anyway. It will happen only rarely when working with big numbers and the false exactness assumption can cause some unexpected problems.

That's why I think only the `Integer` -> `Float` direction is special and needs this warning even when _explicitly casting_.

Reply:
I'm inclined to agree with you, but James has at times state that we should have this warning whenever this conversion happens. Thoughts, @jdunkerley?

On one hand, this warning will occur almost any time this used, which might be too much. On the other hand, I think many users will not be thinking about decimal issues at all; they're likely only going to be dealing with them because they have a database with decimal columns. In which case we might want to err on the side of more warnings.

Reply:
I agree if an explicit cast then the user has chosen to lose precision so don't think we should attach in that case.
I think Integer to float when it exceeds the exact representation is the special case that we should warn about.

Reply:
Done